server/src/main/java/io/crate/execution/dml/upsert/FromRawInsertSource.java

```diff
@@ -21,27 +21,148 @@
+    private final DocTableInfo table;
+    private final TransactionContext txnCtx;
+    private final NodeContext nodeCtx;
+    private final InputFactory inputFactory;
+    private final CheckConstraints<Map<String, Object>, CollectExpression<Map<String, Object>, ?>> checks;
+
+    // holds table columns utilized as well as dynamically added columns of its lifetime
+    private final Map<String, Reference> refLookUpCache = new HashMap<>();
+
+    public FromRawInsertSource(DocTableInfo table,
+                               TransactionContext txnCtx,
+                               NodeContext nodeCtx,
+                               String indexName) {
+        this.table = table;
+        this.inputFactory = new InputFactory(nodeCtx);
+        this.txnCtx = txnCtx;
+        this.nodeCtx = nodeCtx;
+        this.checks = new CheckConstraints<>(
+            txnCtx,
+            inputFactory,
+            new FromSourceRefResolver(table.partitionedByColumns(), indexName),
+            table
+        );
     }
 
     @Override
-    public Map<String, Object> generateSourceAndCheckConstraints(Object[] values) throws IOException {
-        return JsonXContent.JSON_XCONTENT.createParser(
-            NamedXContentRegistry.EMPTY,
-            DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
-            new BytesArray(((String) values[0])).array()
-        ).map();
+    public Map<String, Object> generateSourceAndCheckConstraints(Object[] values) {
+        String src = (String) values[0];
+        Map<String, Object> parsed = Map.of();
+        try {
+            var parser = JsonXContent.JSON_XCONTENT.createParser(
+                NamedXContentRegistry.EMPTY,
+                DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                new BytesArray(src).array());
+            parsed = parser.map();
+            if (parser.getTokenLocation().columnNumber < src.length()) {
+                throw new IllegalArgumentException("failed to parse `" + src.substring(parser.getTokenLocation().columnNumber) + "`");
+            }
+        } catch (IOException e) {
+            if (e instanceof JsonParseException) {
+                throw new IllegalArgumentException("failed to parse");
+            }
+        }
+
+        Map<String, Object> validatedSource = validateRawSource(parsed);
+        mixinDefaults(validatedSource,
+                      // evaluateDefaultValues is invoked per row for expressions like now(), and gen_random_text_uuid().
+                      // Other deterministic expressions like abs(-10) should have been evaluated once and for all by the analyzer.
+                      evaluateDefaultValues(table.defaultExpressionColumns(), txnCtx, nodeCtx));
+
+        GeneratedColumns<Map<String, Object>> generatedColumns = new GeneratedColumns<>(
+            inputFactory,
+            txnCtx,
+            FromSourceRefResolver.WITHOUT_PARTITIONED_BY_REFS,
+            refLookUpCache.values(),
+            table.generatedColumns());
```

Comment:
Is it possible to do this once in the constructor instead of per row?

Reply:
I do not think it is possible, since the 4th parameter represents all present columns for that row at that instant.